public/app/features/explore/state/actions.ts

```diff
@@ -69,10 +70,15 @@ export function addQueryRow(exploreId: ExploreId, index: number): AddQueryRowAct
  * Loads a new datasource identified by the given name.
  */
 export function changeDatasource(exploreId: ExploreId, datasource: string): ThunkResult<void> {
-  return async dispatch => {
-    const instance = await getDatasourceSrv().get(datasource);
-    dispatch(updateDatasourceInstance(exploreId, instance));
-    dispatch(loadDatasource(exploreId, instance));
+  return async (dispatch, getState) => {
+    const newDataSourceInstance = await getDatasourceSrv().get(datasource);
```

Comment:
Since this is async and the next lines get state later, you might end up importing different queries from a different DB.

Reply:
I think you can not. Instance instate is updated later on (L80)